### ISSUE TITLE
HGSS save issue

### ISSUE THREAD

**Issue by Ammako** [kwsch/PKHeX#2697]
**Describe the bug**
Hard to describe with just one line for title, but I've been having an issue from time to time with HGSS and narrowed it down to PKHeX likely being the culprit. What happens is, after editing something in PKHeX (usually something in the PC), the save file loads and plays just fine, but after saving in-game and turning off the game/resetting, the game usually gives a corrupted save file error the next time you try to load the save file again.

Saving multiple times can avoid the problem, but not always. I'm not sure what causes this inconsistency, but I've had it earlier where the error still happened even after saving four times, yet sometimes simply saving twice avoids it. Either way, after the error happens, you can load the previous save file, and after saving again once or twice (you get the "Saving a lot of data..." message), the game fixes itself, and the error doesn't occurs again (until you edit it again in PKHeX, that is.)

Doesn't seem to affect DPPt, only HGSS. Happens the same way on both emu and real hardware (flashcart.) Not sure about retail carts, I guess I could check, but I doubt it would really make a difference given that saving normally works just fine, as long as I didn't edit the save in PKHeX. I also give it 5-10 seconds after saving when on emu or flashcart, before I reset or shut off the game, just in case it needed more time to finish writing to the .sav or something. So I doubt that would be the problem.

**To Reproduce**
1. Load save file in PKHeX
2. Edit stuff in the PC
3. Save the file
4. Load the save in-game
5. Save
6. Soft-reset, or close and re-open the game
7. Save file is corrupted

**Expected behavior**
Being able to use the save file and save in-game after editing in PKHeX, without it resulting in a corrupted save error.

**Screenshots**
https://gfycat.com/insignificantseveralcirriped

**Additional context**
Here's a save where it happens, in case you aren't able to reproduce (there are cases where the error might not happen, I don't know the details though.) Just have to load it and save, then soft-reset once it's done saving.
[Pokémon - HeartGold Version (USA).zip](https://github.com/kwsch/PKHeX/files/4183987/Pokemon.-.HeartGold.Version.USA.zip)

**Comment by kwsch** [comment]
Before saving: Checksums valid.
After saving: Large block checksum is invalid
Reloaded after dismissing "corrupt", and saved:  Checksums valid.

Needs more investigation, but PKHeX is able to recognize something is afoot too 😕

**Comment by kwsch** [comment]
Re-saving the original file in PokeGen doesn't fix any issues, still causes corruption when saved in-game and reloaded.

Gonna assume there's some spooky undocumented behavior, which can't really be resolved now (unless there's a ton of investigation).

**Comment by Ammako** [comment]
I think PKHeX may not be handling box stuff properly for Gen 4 as a whole... and I wonder if this could be related to our problem here.

Take [this Platinum save file](https://github.com/kwsch/PKHeX/files/4565977/Gen.4.-.Pt.Project.493.zip), for example... load it in PKHeX latest stable, clear the first box and save it. Then load the edited save back back up. It ends up loading the backup storage block instead of the current one. Same thing happens on DP and HGSS, too.

**Comment by kwsch** [comment]
I'm assuming that it's something related to the currently-active save block detection. PKHeX bases it off the save-counter that keeps track of how many times the block has been saved; I don't know how the game would be picking the lower-counter block :\

**Comment by Ammako** [comment]
The game picks the correct one, PKHeX doesn't.

People have been complaining for a while from time to time about box edits in Gen 4 not sticking, this may be the culprit.

**Comment by kwsch** [comment]
> clear the first box

https://github.com/kwsch/PKHeX/blame/afd4ff25a5a9c677a8ee263bce4f7fb7b99fd1da/PKHeX.Core/Saves/SAV4.cs#L159

```cs
            if (IsRangeAll(0x00000, 10, 0) || IsRangeAll(0x00000, 10, 0xFF))
                return 1;
```

What is an empty-encrypted-pk4? First 10 bytes are zeroes. An empty b1s1 will cause the logic to use the latter block!

**Comment by md842** [comment]
Hi, I see that this issue is closed, but I got this issue yesterday, so I just wanted to share a workaround that I discovered. Upon getting the corrupted save message, close the game, open the save file in PKHeX, and move a party Pokémon to a different party slot to fix the checksum and allow you back into the game. Export the save back to the game, then open the game. Once you're in the game, you'll need to use the PC in game to move a Pokémon around, which seems to permanently undo whatever PKHeX is doing wrong here (until you use PKHeX to change something in boxes again, then you'll need to repeat this workaround).

**Comment by kwsch** [comment]
@md842 this issue has been fixed since the latest release a few weeks ago. The latest download does not have the issue you describe since the block detection logic has been corrected in the linked commit.

**Comment by md842** [comment]
I see, sorry about that! Turns out my copy is indeed not the latest version.

**Comment by EderParis** [comment]
Just got the same issue today, I'm using PKHex version 20201031

**Comment by Naixatloz** [comment]
This happened to me today. I'm on PKHex version 20221201.

**Comment by KPWave** [comment]
I'm getting this still with SoulSilver; I'm using PKHeX (20230227).

**Comment by Lusamine** [comment]
I am not sure how you people expect anyone to help when you necro a closed issue from 2020 and provide no data, save, or other information.

**Comment by KPWave** [comment]
Sorry, shoulda known better.

I was using PKHeX (like I said latest version atm; 202330227) to add Pokemon to my box; I was using MelonDS on Win11 to run the game. Loading in the game after making the edits is fine and everything starts normally and the edits are there, but always after saving and restarting, I'd get the "The save data is corrupted. Loading previous save file" message during the next boot. Once it loads in after this the changes are still there oddly, and if I save again and reboot everything is fine like usual, the changes remain, and no more corruption screen until the next time I make changes.

Also here attaching 3 saves: save 1 is right after the edits are made, save 2 is after opening the game, saving, and closing it (this is the one that will have the corruption screen), and save 3 is after saving again once the corruption screen has already happened.
[SoulSilver_bug_saves.zip](https://github.com/kwsch/PKHeX/files/11067490/SoulSilver_bug_saves.zip)

**Comment by Naixatloz** [comment]
I no longer have the affected saves, but I can add that in my case, I was deleting Pokemon from my PC box instead of adding any, and I was using an official cartridge. Otherwise it's basically the exact same story.

**Comment by gingrentse** [comment]
Today I had opened the new save and try to test the issue:

1. The new save can edit monsters in PC boxes

2. Save in game

3. No corrupted

4. Use PKHeX > Tools > Data > Load boxes

5. After use the load boxes function, edit monsters in PC boxes and save in game.

6. Found corrupted in new save.

7. Only found this issue in HGSS, PT and White 2 have not.

Is this a bug for Load boxes in HGSS?

**Comment by LBXComputers** [comment]
Getting this often in Soul Silver legitimate cartridge.

Save in game
Export save file with Checkpoint and copy file to PC
Open save file in PKHeX 20240727
Do something in boxes such as re-order all Pokémon across all boxes in PokeDex order
Save the file
Restore back to the cartridge with Checkpoint or import the save to DesMume or use the save in MelonDS
Often, it will give the save corruption error

I've found that double-saving the save file mitigates this mostly.

I've also found that sometime when you use the "delete all from box" option in PKHeX i.e. to empty a box of breedjects, after the save corruption message, some of the Pokémon in that box will be there again or even duplicated.

I think this all started on my save file after I re-ordered the boxes.
